File: mini-era/cv/svm/src/c/takeStep.c
/********************************
Author: Sravanthi Kota Venkata
********************************/

#include "svm.h"

int takeStep(int i, int j, F2D* a, float C, F2D* e, F2D* Y, F2D* X, float eps, float* b, int N, int dim)
{
    int ret=1;
    float s;
    int m, n, k;
    float Ei, Ej, gamma, L, H;
    F2D *a_old;
    float k11, k12, k22, eta;
    F2D *temp, *temp1, *temp2;
    float t, t1, t2;
    float bnew, delta_b;
    float c1, c2, Lobj, Hobj;
    
    if(i==j)
        return 0;
        
    a_old = fDeepCopy(a);
    
    if( asubsref(a_old,i)>0 && asubsref(a_old,i)<C ) 
        Ei = asubsref(e,i);
    else
        Ei = cal_learned_func(i, a, b, N, Y, X, dim) - asubsref(Y,i);

    if( asubsref(a_old,j)>0 && asubsref(a_old,j)<C ) 
        Ej = asubsref(e,j);
    else
        Ej = cal_learned_func(j, a, b, N, Y, X, dim) - asubsref(Y,j);

    s = asubsref(Y,i) * asubsref(Y,j);

    if( asubsref(Y,i) == asubsref(Y,j) )
    {
        gamma = asubsref(a_old,i) + asubsref(a_old,j);
        if(gamma > C)
        {
            L = gamma - C;
            H = C;
        }
        else
        {
            L = 0;
            H = gamma;
        }
    }
    else
    {
        gamma = asubsref(a_old,i) - asubsref(a_old,j);
        if(gamma > 0)
        {
            L = 0;
            H = C - gamma;
        }
        else
        {
            L = -gamma;
            H = C;
        }
    }
    

    if(L==H)
    {
        fFreeHandle(a_old);
        return 0;
    }
           
    temp = fMallocHandle(1, X->width);
    temp1 = fMallocHandle(1, X->width);

    for(m=0; m<X->width; m++)
    {
        asubsref(temp,m) = subsref(X,i,m);
        asubsref(temp1,m) = subsref(X,j,m);
    }

    k11 = polynomial(3, temp, temp, dim);
    k12 = polynomial(3, temp, temp1, dim);
    k22 = polynomial(3, temp1, temp1, dim);
    eta = 2 * k12 - k11 - k22;

    fFreeHandle(temp1);
    fFreeHandle(temp);

    if(eta<0)
    {
        asubsref(a,j) = asubsref(a_old,j) + asubsref(Y,j) * (Ej-Ei)/eta;
        if( asubsref(a,j) < L)
            asubsref(a,j) = L;
        else if( asubsref(a,j) > H )
            asubsref(a,j) = H;
    }
    else
    {
        c1 = eta/2;
        c2 = asubsref(Y,j) * (Ei-Ej) - eta * asubsref(a_old,j);
        Lobj = c1 * L * L + c2 * L;
        Hobj = c1 * H * H + c2 * H;

        if (Lobj > (Hobj+eps))
            asubsref(a,j) = L;
        else if (Lobj < (Hobj-eps))
            asubsref(a,j) = H;
        else
            asubsref(a,j) = asubsref(a_old,j);
    }
    
    if( fabsf( asubsref(a,j)- asubsref(a_old,j) ) < (eps* (asubsref(a,j) + asubsref(a_old,j) +eps)) )
    {
        fFreeHandle(a_old);
        return 0;
    }

    asubsref(a,i) = asubsref(a_old,i) - s * ( asubsref(a,j) - asubsref(a_old,j) );

    if( asubsref(a,i) < 0)
    {
        asubsref(a,j) = asubsref(a,j) + s * asubsref(a,i);
        asubsref(a,i) = 0;
    }
    else if (asubsref(a,i) > C)
    {
        t = asubsref(a,i) - C;
        asubsref(a,j) = asubsref(a,j) + s * t;
        asubsref(a,i) = C;
    }

    /** Update threshold to react change in Lagrange multipliers **/

    if( asubsref(a,i) > 0 && asubsref(a,i) < C )
        bnew = arrayref(b,0) + Ei + asubsref(Y,i) * (asubsref(a,i) - asubsref(a_old,i)) * k11 + asubsref(Y,j) * (asubsref(a,j) - asubsref(a_old,j)) * k12;
    else
    {
        if( asubsref(a,j) > 0 && asubsref(a,j) < C )
            bnew = arrayref(b,0) + Ej + asubsref(Y,i) * (asubsref(a,i) - asubsref(a_old,i)) * k12 + asubsref(Y,j) * (asubsref(a,j) - asubsref(a_old,j)) * k22;
        else
        {
            float b1, b2;
            b1 = arrayref(b,0) + Ei + asubsref(Y,i) * (asubsref(a,i) - asubsref(a_old,i)) * k11 + asubsref(Y,j) * (asubsref(a,j) - asubsref(a_old,j)) * k12;
            b2 = arrayref(b,0) + Ej + asubsref(Y,i) * (asubsref(a,i) - asubsref(a_old,i)) * k12 + asubsref(Y,j) * (asubsref(a,j) - asubsref(a_old,j)) * k22;
            bnew = (b1 + b2) / 2;
        }
    }
    delta_b = bnew - arrayref(b,0);
    arrayref(b,0) = bnew;

    /** Update error cache using new Lagrange multipliers 24ai **/

    t1 = asubsref(Y,i) * (asubsref(a,i)-asubsref(a_old,i));
    t2 = asubsref(Y,j) * (asubsref(a,j)-asubsref(a_old,j));

    temp = fMallocHandle(1, X->width);
    temp1 = fMallocHandle(1, X->width);
    temp2 = fMallocHandle(1, X->width);

    for (k=0; k<N; k++)
    {
        if (0 < asubsref(a_old,i) && asubsref(a_old,i) < C )
        {

            for(m=0; m<X->width; m++)
            {
                asubsref(temp,m) = subsref(X,i,m);
                asubsref(temp1,m) = subsref(X,k,m);
                asubsref(temp2,m) = subsref(X,j,m);
            }

            asubsref(e,k) = asubsref(e,k)+t1 * polynomial(3, temp, temp1, dim) + t2 * polynomial(3, temp2, temp1, dim) - delta_b;
            asubsref(e,i) = 0;
            asubsref(e,j) = 0;
        }
    }
    
    fFreeHandle(a_old);
    fFreeHandle(temp);
    fFreeHandle(temp1);
    fFreeHandle(temp2);
    ret = 1;
    return ret;
}

File: mini-era/cv/svm/src/matlab/takeStep.m
function [a,b,e,ret]=takeStep(i, j, a1, C, e1, Y, X, eps, b1, N, dim)

a = a1;
e = e1;
b = b1;

ret = 1;
if i == j
    ret =0;
end

if( ret ~= 0)
     %variable initialization%
    a_old = a;

    if a_old(i,1) > 0 && a_old(i,1) < C
        Ei = e(i,1);
    else
        Ei = cal_learned_func(i, a, b, N, Y, X, dim) - Y(i,1);
    end

    if a_old(j,1) > 0 && a_old(j,1) < C
        Ej = e(j,1);
    else
        Ej = cal_learned_func(j, a, b, N, Y, X, dim) - Y(j,1);
    end

    s = Y(i,1) * Y(j,1);

    %Compute L, H%
    if Y(i,1) == Y(j,1)
        gamma = a_old(i,1) + a_old(j,1);
        if gamma > C
            L = gamma-C;
            H = C;
        else
            L = 0;
            H = gamma;
        end
    else
        gamma = a_old(i,1) - a_old(j,1);
        if gamma > 0
            L = 0;
            H = C - gamma;
        else
            L = -gamma;
            H = C;
        end
    end
    

    if L == H
        ret=0;
    end
    
end    

if(ret ~=0)
%     %Compute eta
    k11 = polynomial(3,X(i,:), X(i,:), dim);
    k12 = polynomial(3,X(i,:), X(j,:), dim);
    k22 = polynomial(3,X(j,:), X(j,:), dim);
    eta = 2 * k12 - k11 - k22;
    
    if eta < 0
        a(j,1) = a_old(j,1) + Y(j,1) * (Ej - Ei) / eta;
        if a(j,1) < L
            a(j,1) = L;
        elseif a(j,1) > H
            a(j,1) = H;
        end
    else
        %Compute Lobj, Hobj: objective function at a2=L, a2=H 22di
        c1 = eta/2;
        c2 = Y(j,1) * (Ei-Ej)- eta * a_old(j,1);
        Lobj = c1 * L * L + c2 * L;
        Hobj = c1 * H * H + c2 * H;

        if (Lobj > Hobj+eps)
            a(j,1) = L;
        elseif (Lobj < Hobj-eps)
            a(j,1) = H;
        else
            a(j,1) = a_old(j,1);
        end
    end
    
end


if( ret~= 0)

    if abs(a(j,1)-a_old(j,1)) < eps*(a(j,1)+a_old(j,1)+eps)
        ret=0;
    else
        a(i,1) = a_old(i,1) - s * (a(j,1) - a_old(j,1));
        if a(i,1) < 0
            a(j,1) = a(j,1) + s * a(i,1);
            a(i,1) = 0;
        elseif a(i,1) > C
            t = a(i,1)-C;
            a(j,1) = a(j,1) + s * t;
            a(i,1) = C;
        end

        %Update threshold to reect change in Lagrange multipliers
        if a(i,1) > 0 && a(i,1) < C
            bnew = b + Ei + Y(i,1) * (a(i,1) - a_old(i,1)) * k11 + Y(j,1) * (a(j,1) - a_old(j,1)) * k12;
        else
            if a(j,1) > 0 && a(j,1) < C
                bnew = b + Ej + Y(i,1) * (a(i,1) - a_old(i,1)) * k12 + Y(j,1) * (a(j,1) - a_old(j,1)) * k22;
            else
                b1 = b + Ei + Y(i,1) * (a(i,1) - a_old(i,1)) * k11 + Y(j,1) * (a(j,1) - a_old(j,1)) * k12;
                b2 = b + Ej + Y(i,1) * (a(i,1) - a_old(i,1)) * k12 + Y(j,1) * (a(j,1) - a_old(j,1)) * k22;
                bnew = (b1 + b2) / 2;
            end
        end
        delta_b = bnew - b;
        b = bnew;

        %Update error cache using new Lagrange multipliers 24ai

        t1 = Y(i,1) * (a(i,1)-a_old(i,1));
        t2 = Y(j,1) * (a(j,1)-a_old(j,1));
        for k=1:N
            if 0 < a_old(i,1) && a_old(i,1) < C
                e(k,1) = e(k,1)+t1 * polynomial(3,X(i,:),X(k,:),dim) + t2 * polynomial(3,X(j,:),X(k,:),dim) - delta_b;
                e(i,1) = 0;
                e(j,1) = 0;
            end
        end
        ret = 1;
    end
    
end
    
end
